### PR TITLE
Mesh packer V2

### PR DESCRIPTION
**Pull request by Przemog1**

### CODE REVIEW

File: include/nbl/asset/utils/CCPUMeshPackerV1.h [hunk 1]
```diff
@@ -118,14 +73,14 @@ class CCPUMeshPackerV1 final : public IMeshPacker<ICPUMeshBuffer, MDIStructType>
...
+		if(isInstancingEnabled)
+			_NBL_ALIGNED_FREE(m_perInsVtxBuffAlctrResSpc);
 	}
 
 	template <typename MeshBufferIterator>
 	ReservedAllocationMeshBuffers alloc(const MeshBufferIterator mbBegin, const MeshBufferIterator mbEnd);
 
-	//TODO: test
-	void free(const ReservedAllocationMeshBuffers& ramb)  // TODO: this isn't special so do it in the base class
+	void free(const ReservedAllocationMeshBuffers& ramb)
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/utils/CCPUMeshPackerV1.h:83]: huh why not?

File: include/nbl/asset/utils/CCPUMeshPackerV1.h [hunk 2]
```diff
@@ -156,13 +111,18 @@ class CCPUMeshPackerV1 final : public IMeshPacker<ICPUMeshBuffer, MDIStructType>
...
+	virtual void shrinkOutputBuffersSize() override
 	{
-		m_allocParams.MDIDataBuffSupportedCnt = m_MDIDataAlctr.safe_shrink_size(0u, 1u);
-		m_allocParams.perInstanceVertexBuffSupportedByteSize = m_perInsVtxBuffAlctr.safe_shrink_size(0u, 1u);
-		m_allocParams.indexBuffSupportedCnt = m_idxBuffAlctr.safe_shrink_size(0u, 1u);
-		m_allocParams.vertexBuffSupportedByteSize = m_vtxBuffAlctr.safe_shrink_size(0u, 1u);
-		// TODO: SHRINK ACTUAL ALLOCATORS AND THEIR RESERVED SPACES! (CREATE NEW ALLOCATORS WITH NEW RESERVED SPACES, THEN DELETE OLD RESERVED, THEN SWAP ALLOCATOR AND RESERVED MEMBERS) 
+		base_t::shrinkOutputBuffersSize();
+
+		uint32_t mdiDataBuffNewSize = m_perInsVtxBuffAlctr.safe_shrink_size(0u, alctrTraits::max_alignment(m_perInsVtxBuffAlctr));
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/utils/CCPUMeshPackerV1.h:118]: `mdiDataBuffNewSize` is a really bad variable name here!

shouldn't it be `perInstanceDataBuffNewSize` ?

File: include/nbl/asset/utils/CCPUMeshPackerV1.h [hunk 3]
```diff
@@ -304,8 +262,9 @@ typename CCPUMeshPackerV1<MDIStructType>::ReservedAllocationMeshBuffers CCPUMesh
...
 		return invalidReservedAllocationMeshBuffers;
 	}
-	
-	if (m_vtxBuffAlctrResSpc)
+
+	bool arePerVtxAttribsEnabled = alctrTraits::get_total_size(m_vtxBuffAlctr) == 0 ? false : true;
+	if (arePerVtxAttribsEnabled)
 	{
 		vtxAllocAddr = m_vtxBuffAlctr.alloc_addr(vtxCnt * m_vtxSize, 1u);
 		if (vtxAllocAddr == INVALID_ADDRESS)
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/utils/CCPUMeshPackerV1.h:270]: have a look at `SRAIIBasedExiter`, you could wrap up all the 4 failure cases into it

P.S. I might actually do it myself

File: include/nbl/asset/utils/IMeshPacker.h [hunk 4]
```diff
@@ -94,34 +81,28 @@ class IMeshPackerBase : public virtual core::IReferenceCounted
...
 
             if (allocParams.vertexBuffSupportedByteSize)
             {
-                m_vtxBuffAlctrResSpc = _NBL_ALIGNED_MALLOC(core::GeneralpurposeAddressAllocator<uint32_t>::reserved_size(32u, allocParams.vertexBuffSupportedByteSize, allocParams.vertexBufferMinAllocByteSize), _NBL_SIMD_ALIGNMENT);
-                _NBL_DEBUG_BREAK_IF(m_vtxBuffAlctrResSpc == nullptr);
-                assert(m_vtxBuffAlctrResSpc != nullptr);
-                m_vtxBuffAlctr = core::GeneralpurposeAddressAllocator<uint32_t>(m_vtxBuffAlctrResSpc, 0u, 0u, 32u, allocParams.vertexBuffSupportedByteSize, allocParams.vertexBufferMinAllocByteSize);
+                void* resSpcTmp = _NBL_ALIGNED_MALLOC(core::GeneralpurposeAddressAllocator<uint32_t>::reserved_size(32u, allocParams.vertexBuffSupportedByteSize, allocParams.vertexBufferMinAllocByteSize), _NBL_SIMD_ALIGNMENT);
+                assert(resSpcTmp != nullptr);
+                m_vtxBuffAlctr = core::GeneralpurposeAddressAllocator<uint32_t>(resSpcTmp, 0u, 0u, 32u, allocParams.vertexBuffSupportedByteSize, allocParams.vertexBufferMinAllocByteSize);
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/utils/IMeshPacker.h:94]: what does the `32u` mean?

File: include/nbl/asset/utils/IMeshPacker.h [hunk 5]
```diff
@@ -196,19 +215,19 @@ class IMeshPacker : public IMeshPackerBase
     struct IdxBufferParams
     {
         SBufferBinding<ICPUBuffer> idxBuffer;
-        uint32_t idxCnt = 0u; //TODO: if you will be sure that it will not be used anywhere, delete this and modify `convertIdxBufferToTriangles` to return only idx buffer
+        uint32_t idxCnt = 0u;
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/utils/IMeshPacker.h:218]: so where is it being used?

File: include/nbl/asset/utils/IMeshPacker.h [hunk 6]
```diff
@@ -351,16 +370,23 @@ class IMeshPacker : public IMeshPackerBase
...
                     nextTriangle++;
                     batchArea = newBatchArea;
+
+                    if(aabbs != nullptr && (nextTriangle == triangleArrayEnd || (i + 1u) >= m_maxTriangleCountPerMDIData))
+                        *aabbs = core::aabbox3df(core::vector3df(min.x, min.y, min.z), core::vector3df(max.x, max.y, max.z));
                 }
 
                 triangleBatches.ranges.push_back(nextTriangle);
             }
+            aabbs++;
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/utils/IMeshPacker.h:389]: missing an `if (aabbs)`

File: include/nbl/asset/utils/IMeshPacker.h [hunk 7]
```diff
@@ -351,16 +370,23 @@ class IMeshPacker : public IMeshPackerBase
...
+
                     extendAABB(nextTriangle);
                     float newBatchArea = halfAreaAABB();
                     if (newBatchArea > kGrowthLimit* batchArea)
                         break;
                     nextTriangle++;
                     batchArea = newBatchArea;
+
+                    if(aabbs != nullptr && (nextTriangle == triangleArrayEnd || (i + 1u) >= m_maxTriangleCountPerMDIData))
+                        *aabbs = core::aabbox3df(core::vector3df(min.x, min.y, min.z), core::vector3df(max.x, max.y, max.z));
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/utils/IMeshPacker.h:384]: move this to line 386 (and move line 389 there too)
```cpp
if (aabbs)
{
    if (nextTriangle == triangleArrayEnd || i==m_maxTriangleCountPerMDIData)
         *aabbs = core::aabbox3df(core::vector3df(min.x, min.y, min.z), core::vector3df(max.x, max.y, max.z));
    aabbs++;
}
```

otherwise you have a bug and no AABB gets written if `m_minTriangleCountPerMDIData==m_maxTriangleCountPerMDIData` or if the batch has less triangles than `m_minTriangleCountPerMDIData`

File: include/nbl/asset/utils/IMeshPackerV2.h [hunk 8]
```diff
@@ -233,8 +306,6 @@ class IMeshPackerV2 : public IMeshPacker<MeshBufferType,MDIStructType>, public I
             return this->mdiAllocationOffset!=core::GeneralpurposeAddressAllocator<uint32_t>::invalid_address;
         }
     };
-
-    //TODO: if we use SSBO then there is no need for `arrayElement`
     struct VirtualAttribute
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/utils/IMeshPackerV2.h:309]: merge my changes from #109

File: include/nbl/asset/utils/IMeshPackerV2.h [hunk 9]
```diff
@@ -539,34 +610,45 @@ class IMeshPackerV2 : public IMeshPacker<MeshBufferType,MDIStructType>, public I
...
+
+        const auto& idxAlctr = otherMp->getIndexAllocator();
+        alctrBuffSz = alctrTraits::get_total_size(idxAlctr);
+        resSpcTmp = _NBL_ALIGNED_MALLOC(alctrTraits::reserved_size(alctrBuffSz, idxAlctr), _NBL_SIMD_ALIGNMENT);
+        m_idxBuffAlctr = core::GeneralpurposeAddressAllocator<uint32_t>(alctrBuffSz, idxAlctr, resSpcTmp);
+
+        const auto& vtxAlctr = otherMp->getVertexAllocator();
+        alctrBuffSz = alctrTraits::get_total_size(vtxAlctr);
+        resSpcTmp = _NBL_ALIGNED_MALLOC(alctrTraits::reserved_size(alctrBuffSz, vtxAlctr), _NBL_SIMD_ALIGNMENT);
+        m_vtxBuffAlctr = core::GeneralpurposeAddressAllocator<uint32_t>(alctrBuffSz, vtxAlctr, resSpcTmp);
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/utils/IMeshPackerV2.h:645]: I think this is why I suggested an overload of `initializeCommonAllocators` which takes const references to the allocators we're supposed to copy

**Comment by devshgraphicsprogramming** [comment]
Merge `pathtracing1` branch (at working commit ID:) and make example 22 work, then I'll merge this PR into `master`

File: include/nbl/asset/utils/CCPUMeshPackerV1.h (hunk 1 above)
Comment (Przemog1) [reply include/nbl/asset/utils/CCPUMeshPackerV1.h:83]: `CCPUMeshPackerV1::free` differ from `CCPUMeshPackerV2::free`, also `CCPUMeshPackerV1::ReservedAllocationMeshBuffers` and `CCPUMeshPackerV2::ReservedAllocationMeshBuffers` are different types
Shared code could be somehow moved to the base class, but it would be complex and unnecessary imo

File: include/nbl/asset/utils/CCPUMeshPackerV1.h (hunk 2 above)
Comment (Przemog1) [reply include/nbl/asset/utils/CCPUMeshPackerV1.h:118]: oops

File: include/nbl/asset/utils/IMeshPacker.h [hunk 10]
```diff
@@ -214,15 +214,14 @@ class IMeshPacker : public IMeshPackerBase
...
-        uint32_t idxCnt = 0u;
+        SBufferBinding<ICPUBuffer> idxBuffer = { 0u, nullptr };
         E_INDEX_TYPE indexType = EIT_UNKNOWN;
     };
 
     //TODO: functions: constructTriangleBatches, convertIdxBufferToTriangles, deinterleaveAndCopyAttribute and deinterleaveAndCopyPerInstanceAttribute
     //will not work with IGPUMeshBuffer as MeshBufferType, move it to new `ICPUMeshPacker`
 
-    TriangleBatches constructTriangleBatches(const MeshBufferType* meshBuffer, IdxBufferParams idxBufferParams, core::aabbox3df* aabbs) const
+    TriangleBatches constructTriangleBatches(MeshBufferType* meshBuffer, IdxBufferParams idxBufferParams, core::aabbox3df* aabbs) const
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/utils/IMeshPacker.h:224]: why did you remove const?

File: include/nbl/asset/utils/IMeshPacker.h (hunk 4 above)
Comment (devshgraphicsprogramming) [reply include/nbl/asset/utils/IMeshPacker.h:94]: ah probably because biggest attribute you can ever hope for is `dvec4` and the offsets would need alignment to `sizeof(dvec4)`

File: include/nbl/asset/utils/IMeshPacker.h (hunk 10 above)
Comment (Przemog1) [reply include/nbl/asset/utils/IMeshPacker.h:224]: https://github.com/Devsh-Graphics-Programming/Nabla/pull/111/files/3e94ece32799e431842f15e4768741d7dbfa95b8..16f6abdb8bf00336e2b0da3c9cbcb4df8cbf64fe#diff-9da5291f7aea536f7b8a62660d1bf42cc8b70620472e0b385f6775a789e4d636R276-R282

File: include/nbl/asset/utils/IMeshPacker.h [hunk 11]
```diff
@@ -275,15 +273,13 @@ class IMeshPacker : public IMeshPackerBase
...
-            //    wasTmpIdxBufferSet = true;
-            //    meshBuffer->setIndexBufferBinding(idxBufferParams.idxBuffer);
-            //}
+            const bool wasTmpIdxBufferSet = idxBufferParams.idxBuffer.buffer.get() != nullptr ? true : false;
+            if (wasTmpIdxBufferSet)
+            {
+                auto tmpBuffer = std::move(meshBuffer->getIndexBufferBinding());
+                meshBuffer->setIndexBufferBinding(std::move(idxBufferParams.idxBuffer));
+                idxBufferParams.idxBuffer = std::move(tmpBuffer);
+            }
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/utils/IMeshPacker.h:282]: instead of screwing around with the index buffer pinned to the `meshBuffer` maybe sidestep the
`auto triangleIndices = IMeshManipulator::getTriangleIndices(meshBuffer, ix++);`
with your own implementation that uses the new index buffer ?

Comment (Przemog1) [reply include/nbl/asset/utils/IMeshPacker.h:282]: ok

File: include/nbl/asset/utils/CCPUMeshPackerV1.h (hunk 1 above)
Comment (devshgraphicsprogramming) [reply include/nbl/asset/utils/CCPUMeshPackerV1.h:79]: did it myself.

File: include/nbl/asset/utils/CCPUMeshPackerV1.h (hunk 3 above)
Comment (devshgraphicsprogramming) [reply include/nbl/asset/utils/CCPUMeshPackerV1.h:260]: if it works, it works, I'm leaving it alone
